### PR TITLE
Support for "multi-profile" UKIs (i.e. multiple alternative cmdlines per UKI)

### PR DESCRIPTION
**Pull request by poettering**
This is a WIP thing for now, comments welcome.

The interesting bits are in the last commit's message.

Before this is merged this needs to support in ukify (to generate images like this) and sd-boot (to synthesize multiple menu entries from a single UKI)

Fixes: #24539

### CODE REVIEW

**Comment by bluca** [comment]
> 1. An UKI's command line can be prefixed with an argument such as "@1" or
>  "@2" or "https://github.com/3" which indicates the "profile" to boot
>
> 2. The stub will first search for its usual set of PE sections
>    (hereafter called "base sections"). If a profile other than @0 is
>    specified then it will suffix all known UKI PE section names with the
>    profile number, and will look for sections named that way. Each such
>    section that exists will override the matching base section

This seems all unnecessarily convoluted and complex to me, and very hard to understand. It would be much better to just make it simpler: no need for prefixes or suffixes, just have multiple .cmdline sections - the PE spec allows it anyway. Then the selector can simply show the options in the order they are defined in the PE header. No need to truncate, no need for special characters that are used nowhere else, no need for special logic, no need for an entire new concept of "profile". Just append as many sections as one wants to have, and that's it.

**Comment by poettering** [comment]
> This seems all unnecessarily convoluted and complex to me, and very hard to understand. It would be much better to just make it simpler: no need for prefixes or suffixes, just have 

We only have suffixes?

> multiple .cmdline sections - the PE spec allows it anyway. 

Uh? That is not remotely user friendly. The boot loader should offer choices between "Factory Reset", "Boot into Storage Target Mode", "Rescue Mode", and so on. Not between some cryptic kernel command line, that don't give you a slightest hint what they are about unless you have a phd in Linuxology. i.e. "foooxyz=bar zzz=abc" is not friendly UI.

Also, for .dtb and .ucode what are you going to even show in the menu? 25 times the same string ".dtb"?

Also, this is a hyperprivileged environment, I purposefully only allow curated combinations, because it shouldn't be allowed to necessarily combine things too liberally in ways that don't work together. Moreover we already have something like that via add-ons, and it scales much better if you have many options.

Hence, I am pretty sure we have quite different usecases here. I don't really understand yours though, but I certainly udnerstand mine: to bring a few curated options into the boot menu, and not to imply a complex UI which allows selection gazillions of combinations. I actually fail to see at all, where something like that would be even useful.

I think curation is necessary here, and sd-boot so far was about providing "managed" menu entries, that show high-level, predefined sets of configs (that is quite literally what type #1 boot loader spec stuff is all about after all: binding different choices together and turning them into menu items). This here is basically a way to make some of the power of type #1 available in type #2 so that we can properly sign/measure the combinations and allow them in SB lockdown mode, where the combinations of type #1 are not allowed.

> Then the selector can simply show the options in the order they are defined in the PE header. 

I don't think that is desirable nor realistic the way you propose: if you have no useful description to show, just always the same string ".dtb" for 33 items, then that's just garbage UI. Moreover you no longer curate entries, you don't give users a hand what combinations actually make sense, or can sign off on specific combinations that are good and useful.

In my proposal you can curate entries, give them individual menu entry strings to show in the menu, and then we strictly enforce that only the curated, vetted items can actually be booted.

> No need to truncate, no need for special characters that are used nowhere else, no need for special logic, no need for an entire new concept of "profile". Just append as many sections as one wants to have, and that's it.

That sounds pretty unnecessary to me. I simply don't see that usecase at all, and at the very least it's not the usecase I have in mind (which is purely just having curated stuff for "Boot Into Storage Mode", and "Factory Reset" as mentioned above).

Moreover, I would be careful with defining rules that multiple sections of the same name should always mean XOR (i.e. choose one, consider them alternatives), and expose them as separate user-facing options you have to pick one from. For things like .dtb people want automatic matching, and combination apparently. And for stuff like .initrd it definitely might make sense one day to allow combination of stuff. (i.e. on profile 1, combine initrd X and Y but in profile 2 combine X and Z instead). Moreover, you might even want to allow selected 0 entries of the available ones. 

Hence, I think I know exactly what *I* want to deliver with this, but I have trouble to undestand what you are trying to cover with your approach, and I don't see how it could even fly, and it would lock us down under a definition of things that is not what I think people might actually want.

**Comment by poettering** [comment]
I changed the approach a bit now. There's now a concept of ".profile" PE sections which can appear multiple times, and which each introduce a new profile. This section can carry some metadata (such as textual identifier string for the profile, and a human readable string), too.

Seems to work fine here.

This isn't ready fore review yet. This adds the necessary changes to sd-stub, sd-boot, systemd-measure and ukify. This doesn't cover yet updates for bootctl/bootspec.c in userspace to be able to process these multi-profile UKIs yet. Also no coverage by pcrlock yet. And no docs

**Comment by Itxaka** [comment]
Couple of things that are not clear to me.

What can be overriden in each `.profile` ? Everything? Even the kernel?

Supposedly sd-boot will read the whole thing and show entries for each one? Will this work only via autodiscovery of efi files or also if we set a conf pointing to that file?


```
An UKI's command line can be prefixed with an argument such as "@0" or
   "@1" or "@2" which indicates the "profile" to boot.
```

I dont get this, where do you put that @ ? in the `.cmdline` section? In the config? What good is that? How is that used? Very unclear to me. Is this for non-SecureBoot scenarios in which you can override the cmdline manually?

Also it looks like the `.cmdline` are additive to the base one. How would that work with stanzas that dont get the latest value? IIRC some cmdline values if set several times, are not picked up by dracut correctly and the last value set does not won. Is that something we have to consider? i.e. setting something in the base profile cmdline can clash with other cmdlines?

On measurements, how would this work? What measurements would the UKI file have? All possible variations? i.e. base sections+profile0, base sections+profile1, etc...?

```
Or in other words, if you boot
   profile `@3`, then we'll measure .cmdline (assuming it exists) of
   profile 3, and *not* measure .cmdline of the base profile
```

This does not make sense to me. If what we are booting is base profile + profile 3 appended, we should measure the whole thing, i.e. what ends up in /proc/cmdline. Otherwise if the base profile is altered or manipulated it will get run, but the measurements will still match, which breaks the promise of the measured boot no?

**Comment by poettering** [comment]
> Couple of things that are not clear to me.
> 
> What can be overriden in each `.profile` ? Everything? Even the kernel?

Well, yes, if you like. At least technically this is allowed. But of course IRL that makes little sense to do. I'd assume that profiles mostly will override .cmdline, and potentially .dtb or so, but not much else.

> Supposedly sd-boot will read the whole thing and show entries for each one? Will this work only via autodiscovery of efi files or also if we set a conf pointing to that file?

Hmm, not sure I grok the question, but are you asking about what happens if a type 1 boot loader spec entry references a UKI? We would not automatically synthesize multiple entries from that. Multiple menu entries are only synthesized from UKis dropped into EFI/Linux/, i.e. type 2 boot loader spec entries.

You can however invoke a specific profile of a UKI from a type 1 entry: just do "options @4" in your type 1 entry, and reference some UKI via "kernel" and it the entry will boot into profile 4 of the UKI.

> I dont get this, where do you put that @ ? in the `.cmdline` section? In the config? What good is that? How is that used? Very unclear to me. Is this for non-SecureBoot scenarios in which you can override the cmdline manually?

The UKI expects this on its own command line. i.e. sd-boot must tell the UKI somehow which profile to use, hence it will pass `@4` (or so) as first argument to the the UKI when invoking it. If you use a UKI without sd-boot (for example, because you register it directly in the BootXYZ efi vars) then you encode the option to pass there too. 

systemd-stub will look for the `@xyz` argument, then select the PE sections matching the specified profile and boot into that. If you do not specify any profile via `@xyz`, then we'll boot into profile 0.

Note that `@xyz` is something systemd-stub interprets only when it appears on its own command line. It is *not* considered part of the kernel command line, and systemd-stub will strip it before it uses the rest of its own command line to any invoked kernel.

> Also it looks like the `.cmdline` are additive to the base one. 

Not sure I grok what you mean by that. But note that any PE section contained in a profile will *override* the section of the same name in the base profile. Thus, no, there's no "additive" behaviour of kernel command lines here.

> On measurements, how would this work? What measurements would the UKI file have? All possible variations? i.e. base sections+profile0, base sections+profile1, etc...?

No, we'd mesure the sections actually selected by the profile. i.e. the non-overriden PE sections from the base profile, and the PE sections from the selected profile.

Moreover we'll explicitly, separately measure the number of the selected profile.

**Comment by poettering** [comment]
> Awesome. Do we expect this to be limited at any point to restrict entries or not?

i don't see why we would. UKIs are signed, measured hyper-privileged constructs, no need to limit things there.

> So setting `rd.neednet=1` in base profile cmdline and then `rd.neednet=0` in cmdline for profile 1 will results in the final cmdline when running profile 1 of `rd.neednet=1 rd.neednet=0` ? Or I am misunderstanding the commit comment below?

Yes, you are: what I meant is that the person who puts together the UKI already suffixes the regular cmdline like that. it's not sd-stub that would do that. sd-stub only sees the already suffixed cmdlines of which it picks one

**Comment by poettering** [comment]
> As far as I understood the measurements include the full uki file. With different profiles possible, the measurements will be different for each profile. How would that fit into there and the current existing tools like cryptsetup/cryptenroll ? Does it mean that this change will be a breaking change (i.e. uki generated with systemd 257 with profiles will not unlock if booted with stub 256) as the measurements or the pcrsign format will be different?

this should just work with this PR.

You invoke systemd-measure on a specific combiation of sections which then generates the signature for you for that combination. If we have multiple profiles we can just invoke systemd-measure multiple times, once for each profile, each time specifying the combination we want to cover. THis in fact is how the ukify changes that are part of this PR already do it: you take an existing UKI, then you extend it with a new profile, and sign that new profile (taking the base profile of the existing UKI into consideration) with systemd-measure.

So when you invoke profile `@7` of a specific UKI, then you'd get .prcpkey and .pcrsig matching that profile. And because you enroll only a public key into LUKS it would just work.

(I think in the long run we want to teach ukify that it can build multi-profile UKIs in one go, rather than calling it once for each profile always extending the UKI from the previous step, but this was above my python-fu for now, and the --extend= thing sounded nice anyway)

File: src/boot/efi/pe.c [hunk 1]
```diff
@@ -381,30 +381,216 @@
         if ((size_t) pe.FileHeader.NumberOfSections > SIZE_MAX / sizeof(PeSectionHeader))
                 return EFI_OUT_OF_RESOURCES;
         REENABLE_WARNING;
-        section_table_len = (size_t) pe.FileHeader.NumberOfSections * sizeof(PeSectionHeader);
-        if (section_table_len > SECTION_TABLE_BYTES_MAX)
+        size_t n_section_table = (size_t) pe.FileHeader.NumberOfSections;
+        if (n_section_table * sizeof(PeSectionHeader) > SECTION_TABLE_BYTES_MAX)
```
[the remark this answers is not in the document]
Comment (poettering) [reply src/boot/efi/pe.c]: as usual, size_t might have different sizes on different archs

**Comment by poettering** [comment]
> Just an idea, but if we practically just want to have profiles for `.cmdline` sections, then why not having multiple addons for each profile? Ie having either 1 big addon with multiple profiles in it, or multiple addons under `$BOOT/EFI/Linux/${UKI_NAME}.efi.extra.d/profile@xyz/something.addon.efi` or simply just leave a `profile@xyz.addon.efi` in `${UKI_NAME}.efi.extra.d`?
> I am not sure about the measurement implications though.

I think the point of addons as separate files is twofold:

1. allowing some form of modularity to cut down disk space, security exposure and so on
2. allowing different ownership of resources, i.e. UKI from OS vendor, plus add-on from local admin

But the multi profile stuff isn#t really about this. It's neither about disk space/security exposure, not about ownership of resources. Instead, it's about defining multiple vetted, blessed profiles for booting, that only differ in minor ways from each other. The owner of all profiles is the same (OS vendor), and the disk space issue is neglible. 

In general, one of the primary ideas of UKIs is that they are *unified*, i.e. everything in one file. That's great for security, but also great for robustness (ESPs are VFAT after all), and simplicity. 

Addons is a contradiction to the *unified* approach of UKIs of course, and I think it's fine, if we have really good reasons for this (i.e. coarse modulatory to cover for disk space issues,  different ownership of resources), but unless there's one of these strong reasons to have things split out I am very sure it should always be kept inside the UKI.

**Comment by poettering** [comment]
I have now prepped an extension for the UKI spec formalizing the multi-profile UKI concept a bit:

https://github.com/uapi-group/specifications/pull/108

File: test/units/TEST-86-MULTI-PROFILE-UKI.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    echo "EXTENDED UKI:"
+    ukify inspect /tmp/extended2.efi
+
+    mv /tmp/extended2.efi "$CURRENT_UKI"
+    reboot
+    exit 0
+else
+    # shellcheck disable=SC1090
+    . /run/systemd/stub/profile
```
[the remark this answers is not in the document]
Comment (poettering) [reply test/units/TEST-86-MULTI-PROFILE-UKI.sh:57]: obsolete

**Comment by poettering** [comment]
This is ready for review now, btw. PTAL!

File: src/boot/efi/pe.c [hunk 3]
```diff
@@ -388,30 +386,150 @@ EFI_STATUS pe_file_locate_sections(
...
+        unsigned current_profile = UINT_MAX;
+        for (const PeSectionHeader *p = section_table; p < section_table + n_section_table; p++) {
+
+                if (!pe_section_name_equal((const char*) p->Name, ".profile"))
+                        continue;
+
+                if (current_profile == UINT_MAX)
+                        current_profile = 0;
+                else
+                        current_profile ++;
```
Comment (daandemeyer) [inline src/boot/efi/pe.c]: Spurious whitespace

File: man/systemd-stub.xml [hunk 4]
```diff
@@ -53,13 +53,14 @@
     <para>The UEFI boot stub looks for various resources for the kernel invocation inside the UEFI PE binary
     itself. This allows combining various resources inside a single PE binary image (usually called "Unified
     Kernel Image", or "UKI" for short), which may then be signed via UEFI SecureBoot as a whole, covering all
-    individual resources at once. Specifically it may include:</para>
+    individual resources at once. Specifically it may include include the following PE sections:</para>
```
Comment (daandemeyer) [inline man/systemd-stub.xml]: "include include"?

File: man/systemd-stub.xml [hunk 5]
```diff
@@ -106,10 +115,13 @@
...
 
-    <para>The contents of eight of these nine sections are measured into TPM PCR 11. It is otherwise not used
-    and thus the result can be pre-calculated without too much effort. The <literal>.pcrsig</literal> section
-    is not included in this PCR measurement, since it is supposed to contain signatures for the output of the
-    measurement operation, and thus cannot also be input to it.</para>
+    <para>The contents of 11 of these 12 sections are measured into TPM PCR 11. It is otherwise not used and
+    thus the result can be pre-calculated without too much effort. The <literal>.pcrsig</literal> section is
+    not included in this PCR measurement, since it is supposed to contain signatures for the output of the
+    measurement operation, and thus cannot also be input to it. If an UKI contains multiple profiles, only
+    the PE sections of the selected profile are measured.</para>
```
Comment (daandemeyer) [inline man/systemd-stub.xml]: "only the PE sections of the selected profile are measured."

This sounds wrong, the base sections that are not overridden are also measured.

File: man/systemd-stub.xml [hunk 6]
```diff
@@ -231,19 +243,125 @@
...
 
+  <refsect1>
+    <title>Multi-Profile UKIs</title>
+
+    <para>In many contexts it is useful to allow invocation of a single UKI in multiple different modes (or
+    "profiles") without compromising the cryptographic integrity, measurements and so on of the boot
+    process. For example, a single UKI might provide three distinct profiles: a regular boot one, one that
+    invokes a "factory reset" operation, and one that boots into a storage target mode (as in
+    <citerefentry><refentrytitle>systemd-storagetm.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>). Each
+    profile would then use the same <literal>.linux</literal>, <literal>.initrd</literal> sections, but would
```
Comment (daandemeyer) [inline man/systemd-stub.xml]: "," => "and"

File: man/systemd-stub.xml [hunk 7]
```diff
@@ -231,19 +243,125 @@
...
+    pair, and once after the last appearance of <literal>.profile</literal>. The sections listed before the
+    first <literal>.profile</literal> are considered the "base" profile of the UKI. Each
+    <literal>.profile</literal> section then introduces a new profile, which are numbered starting from
+    zero. The PE sections following each <literal>.profile</literal> are specific to that profile. When
+    booting into a specific profile the base section's profiles are used in combination with the specific
+    profile's sections: if the same section is defined in both, the per-profile section overrides the base
+    profile's version, otherwise the per-profile sections is used together with the base profile sections.</para>
+
+    <para>A UKI that contains no <literal>.profile</literal> is consider equivalent to one that just contains
+    a single <literal>.profile</literal>, as having only single profile @0.</para>
```
Comment (daandemeyer) [inline man/systemd-stub.xml]: having only "a" single profile

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
It's too early for reviews, I'll continue later

File: man/systemd-stub.xml (hunk 5 above)
Comment (poettering) [reply man/systemd-stub.xml]: they are considered part of the profile.. but yeah, i guess i can clarify that.

**Comment by poettering** [comment]
Closing, since all commits of this PR have now been merged separately.

**Pull request closed without merging**